sql-to-dbsp-compiler/lib/sqllib/src/variant.rs

```diff
@@ -416,3 +744,267 @@ pub fn typeofN(value: Option<Variant>) -> String {
+                    Variant::Decimal(Decimal::from(12345)),
+                ),
+                (
+                    Variant::String("f".to_string()),
+                    Variant::Double(F64::new(0.00123)),
+                ),
+                (
+                    Variant::String("d".to_string()),
+                    Variant::Decimal(Decimal::from_str("123.45").unwrap()),
+                ),
+                (
+                    Variant::String("s".to_string()),
+                    Variant::String("foo\nbar".to_string()),
+                ),
+                (Variant::String("n".to_string()), Variant::VariantNull),
+                (
+                    Variant::String("nested".to_string()),
+                    Variant::Map(
+                        [
+                            (
+                                Variant::String("arr".to_string()),
+                                Variant::Array(vec![
+                                    Variant::Decimal(Decimal::from(1)),
+                                    Variant::String("foo".to_string()),
+                                    Variant::VariantNull,
+                                ]),
+                            ),
+                            (
+                                Variant::String("ts".to_string()),
+                                Variant::Timestamp(Timestamp::from_dateTime(
+                                    DateTime::parse_from_rfc3339("2024-12-19T16:39:57Z")
+                                        .unwrap()
+                                        .to_utc(),
+                                )),
+                            ),
+                            (
+                                Variant::String("d".to_string()),
+                                Variant::Date(Date::from_date(
+                                    NaiveDate::from_ymd_opt(2024, 01, 01).unwrap(),
+                                )),
+                            ),
+                            (
+                                Variant::String("t".to_string()),
+                                Variant::Time(Time::from_time(
+                                    NaiveTime::from_hms_opt(17, 30, 40).unwrap(),
+                                )),
+                            ),
+                        ]
+                        .into_iter()
+                        .collect(),
+                    ),
+                ),
+            ]
+            .into_iter()
+            .collect(),
+        );
+
+        let s = serde_json::to_string(&v).unwrap();
+
+        let expected = serde_json::from_str::<serde_json::Value>(
+            r#"{
+                "b": true,
+                "i": 12345,
+                "f": 0.00123,
+                "d": 123.45,
+                "s": "foo\nbar",
+                "n": null,
+                "nested": {
+                    "arr": [1, "foo", null],
+                    "ts": "2024-12-19T16:39:57+00:00",
```

Comment:
json_unparse does not do this for timestamps, perhaps it could

Reply:
it should do it now

Reply:
dates and times too?